Add basic PR testing

scottwittenburg opened the pull request.

scottwittenburg commented:
@burlen @patrickoleary

I think this is ready for review.  Once it looks good, we'll need to work together to change the target over to Kitware/sensei, which will require a few additional steps (e.g. setting up a robot user to post CDash status back to PR, etc.)

We also probably want to address the test failures, so that when it's merged we start off with all green check marks.  I may be able to fix the issues affecting MacOS, but CircleCI (ubuntu) is showing some errors that should probably be looked at by an actual sensei developer.

burlen commented on ci_scripts/ci/circle/run.sh:
```diff
@@ -0,0 +1,102 @@
+API_BASE="https://api.github.com/repos/scottwittenburg/sensei"
+COMMIT=${CIRCLE_SHA1}
+CDASH_STATUS_CONTEXT="cdash"
+PYTHON_EXECUTABLE=python
```
should be python3

burlen commented on ci_scripts/ci/cmake/circle_ubuntu18.04.cmake:
```diff
@@ -0,0 +1,49 @@
+
+macro(dashboard_hook_test)
+  message("dashboard_hook_test")
+  message("Attempting to load module: sensei/3.0.0-catalyst-shared-ci")
```
the name of the file should indicate that it is a catalyst specific file

burlen commented on ci_scripts/ci/gh-actions/macos-setup.sh:
```diff
@@ -0,0 +1,38 @@
+esac
+
+echo "Installing CMake Nightly"
+curl -L https://cmake.org/files/dev/cmake-3.16.20191218-g8262562-Darwin-x86_64.tar.gz | sudo tar -C /Applications --strip-components=1 -xzv
```
could we not use an officially released version of cmake? it seems as though this create an unrealistic scenario. none of our users will be using this version. Other than you none of us on the team will either

burlen commented on ci_scripts/ci/gh-actions/run.sh:
```diff
@@ -0,0 +1,81 @@
+  # https://github.com/open-mpi/ompi/issues/6518
+  export OMPI_MCA_btl=self,tcp
+
+  # Enable overscription in OpenMPI
```
we definitely don't want to run things over subscribed

burlen commented on ci_scripts/dashboard/EnvironmentModules.cmake:
```diff
@@ -0,0 +1,207 @@
+# Distributed under the OSI-approved BSD 3-Clause License.  See accompanying
```
this file would presumably be distributed with cmake

scottwittenburg replied:
This is not distributed with CMake, but you mentioned last week that you don't want to use this anyway, correct?  Along those lines, I just want to be clear what it is you do want:  

1. Do you want us to provide a shell script instead that invokes the `module load ...` command (leaving the environment module installed in the Docker image) so CTest can find the tests to run and the needed libraries?

2. Or did you want to get rid of the environment modules from the Docker image completely and just update the `PATH`, `PYTHONPATH`, and `LD_LIBRARY_PATH` manually before we run the tests?

With either approach above, we get rid of the `EnvironmentModules.cmake` file.

burlen commented on ci_scripts/dashboard/common.cmake:
```diff
@@ -0,0 +1,500 @@
+#   set(ENV{FC}  /path/to/fc)   # Fortran compiler (optional)
+#   set(ENV{LD_LIBRARY_PATH} /path/to/vendor/lib) # (if necessary)
+
+cmake_minimum_required(VERSION 2.8.2 FATAL_ERROR)
```
version no seem low to me

burlen commented on ci_scripts/ci/gh-actions/macos-setup.sh:
```diff
@@ -0,0 +1,38 @@
+#!/bin/bash
+
+echo "Setting up default XCode version"
+case "$GH_YML_JOBNAME" in
```
can we simply use the latest xcode here?

burlen commented on ci_scripts/dashboard/sensei_common.cmake:
```diff
@@ -0,0 +1,89 @@
+#------------------------------------------------------------------------------#
+# Distributed under the OSI-approved Apache License, Version 2.0.  See
+# accompanying file Copyright.txt for details.
+#------------------------------------------------------------------------------#
```
this is not our license

burlen commented on ci_scripts/dashboard/sensei_common.cmake:
```diff
@@ -0,0 +1,89 @@
+# accompanying file Copyright.txt for details.
+#------------------------------------------------------------------------------#
+#
+# ADIOS Common Dashboard Script
```
this is not the adios project

burlen commented on ci_scripts/dashboard/common.cmake:
```diff
@@ -0,0 +1,500 @@
+#------------------------------------------------------------------------------#
+# Distributed under the OSI-approved Apache License, Version 2.0.  See
+# accompanying file Copyright.txt for details.
+#------------------------------------------------------------------------------#
```
this is not our license

scottwittenburg replied:
I see your license now, thanks.  Since I have modified all these scripts, and since the apache 2 license does not require a derivative work of the software, or modifications to the original, to be distributed using the same license, I'm just going to remove these two license statements.  I think this just results in re-distribution under the license in the root directory `LICENSE` file.

burlen commented on ci_scripts/ci/travis/run.sh:
```diff
@@ -0,0 +1,59 @@
+${CTEST} --version
+
+# Define dashboard steps
+declare -a steps=("update" "configure" "build" "test")
```
why run this in a loop? it seems to me to add complexity vs the usual approach

scottwittenburg replied:
Is your question about why it's done in steps?  Or just about the use of a bash loop instead of 4 separate invocations?

burlen commented on .github/workflows/build-and-test.yml:
```diff
@@ -0,0 +1,40 @@
+  pull_request:
+    branches:
+    - master
+    - develop
```
should `release*` be here as well?

scottwittenburg replied:
This should probably be only branches your pull requests are going to target.  If you plan to make pull requests against release branches, then `release*` should be here as well.

burlen commented on .travis.yml:
```diff
@@ -0,0 +1,15 @@
+install:
+  - pip3 install mpi4py
+script:
+  - git reset --hard ${TRAVIS_PULL_REQUEST_SHA}
```
does Travis CI not do the clone & checkout the right branch/commit for us??

scottwittenburg replied:
It first does a merge of the PR branch with master, resulting in CDash showing some random SHA in its update column (one that doesn't match all the other rows).

burlen commented on CTestConfig.cmake:
```diff
@@ -1,6 +1,6 @@
-set(CTEST_PROJECT_NAME "SENSEI")
+set(CTEST_PROJECT_NAME "sensei")
```
`SENSEI ` should be capitalized

burlen commented on ci_scripts/ci/circle/run.sh:
```diff
@@ -0,0 +1,102 @@
+  "description": "Build and test results available on CDash",
+  "context": "${CDASH_STATUS_CONTEXT}"
+}
+EOF
```
I think these "here doc" implement the badge feature that you mentioned are in a future release of cdash. would you please add some comments explaining what these functions do and in which versions they become obsolete and if practical the steps to replace with the official feature?

scottwittenburg replied:
I'll add some explanation of what this is doing and how it can be replaced when next there is a CDash release and NERSC upgrades to use it.  Though to be clear, this isn't about badges, but just about posting the link to the CDash build results in the PR status checks.

scottwittenburg commented:
@burlen  A few updates:

1. I have changed the target of this PR to `develop` which I pushed to my fork from gitlab.
2. I have removed the code that posts a link to CDash build results.
3. Each CI provider gives the ability to configure when CI is run, and at the moment a PR to any branch will trigger CI on all three providers, and any subsequent push to a branch with an open PR will re-trigger the CI to run on all three providers.

FYI @patrickoleary
